tests/lib/resource_test.py

```diff
@@ -317,7 +332,85 @@ def test_test_entrypoint_success():
     event = resource.test_entrypoint.__wrapped__(  # pylint: disable=no-member
         resource, payload, None
     )
-    assert event is sentinel.response
+    assert event is progress_event
 
     mock_model._deserialize.assert_has_calls([call(None), call(None)])
     mock_handler.assert_called_once()
+
+
+def test_schedule_reinvocation_not_in_progress():
+    progress = ProgressEvent(status=OperationStatus.SUCCESS)
+    with patch(
+        "cloudformation_cli_python_lib.resource.boto3.Session", autospec=True
+    ) as mock_session, patch(
+        "cloudformation_cli_python_lib.resource.CloudWatchScheduler", autospec=True
+    ) as mock_scheduler:
+        reinvoke = Resource.schedule_reinvocation(
+            sentinel.request, progress, sentinel.context, sentinel.credentials
+        )
+    assert reinvoke is False
+    mock_session.assert_not_called()
+    mock_scheduler.assert_not_called()
+
+
+def test_schedule_reinvocation_local_callback():
+    progress = ProgressEvent(status=OperationStatus.IN_PROGRESS, callbackDelaySeconds=5)
+    mock_request = Mock(
+        "cloudformation_cli_python_lib.interface.HandlerRequest", autospec=True
+    )()
+    mock_request.requestContext = {}
+    mock_context = Mock(
+        "cloudformation_cli_python_lib.interface.LambdaContext", autospec=True
+    )()
+    mock_context.get_remaining_time_in_millis.return_value = 600000
+    with patch(
+        "cloudformation_cli_python_lib.resource.sleep", autospec=True
+    ) as mock_sleep:
+        reinvoke = Resource.schedule_reinvocation(
+            mock_request, progress, mock_context, sentinel.credentials
+        )
+    assert reinvoke is True
+    mock_sleep.assert_called_once_with(5)
+    assert mock_request.requestContext.get("invocation") == 1
+
+
+def test_schedule_reinvocation_cloudwatch_callback():
```

Comment:
there are a lot of mocks here. i vastly prefer some decoupling, such as creating the platform session somewhere else

Reply:
reduced the mock count somewhat, session is created elsewhere, still quite a few mocks left though